File: 2ºAÑO/1ºCuatrimestre/PDOO/examen_p1_p2/examen/lib/civitas_juego.rb
# encoding:utf-8

require_relative "jugador.rb"
require_relative "dado.rb"
require_relative "mazo_sorpresas.rb"
require_relative "gestor_estados.rb"
require_relative "tablero.rb"

module Civitas
  class Civitas_juego
    
    def initialize(nombres)
      @tablero
      @estado
      
      @jugadores = Array.new
      
      for i in nombres do
        @jugadores << Jugador.new(i)
      end
      
      @gestor_estados = Gestor_estados.new
      @estado = @gestor_estados.estado_inicial
      @indice_jugador_actual = Dado.instance.quien_empieza(@jugadores.size)
      @mazo = Mazo_sorpresas.new
      inicializar_tablero(@mazo)
      inicializar_mazo_sorpresas(@tablero)
    end
    
    
    
=begin
def construir_casa(ip)
      
    end
    
    def construir_hotel(ip)
      
    end

 def cancelar_hipoteca(ip)
      return self.jugador_actual.hipotecar(ip)
    end

=end
    
    def contabilizar_pasos_por_salida(jugador_actual)
      
      seguir = true
      
      while seguir do
        if @tablero.por_salida == 0
          seguir = false
        else
          jugador_actual.pasa_por_salida
        end
      end
      
    end
    
    def final_del_juego
      final = false
      
      for i in @jugadores do
        if i.en_bancarrota
          final = true
        end
      end
      final
    end
    
    def jugador_actual
      return @jugadores[@indice_jugador_actual]
    end
    
    def casilla_actual
      casilla = @jugadores.indice_jugador_actual.num_casilla_actual
      return @tablero.get_casilla(casilla)
    end
    
=begin
def hipotecar(ip)
      return self.jugador_actual.hipotecar(ip)
    end
=end
    
    
    def info_jugador_texto
      @jugadores[@indice_jugador_actual].to_s
    end
    
    def inicializar_mazo_sorpresas(tablero)
        @mazo.al_mazo(Sorpresa.new_generico(Tipo_sorpresa::PAGARCOBRAR, -100, "Te han multado por ir a 10km/h con el coche TORTUGA"))
        @mazo.al_mazo(Sorpresa.new_generico(Tipo_sorpresa::PAGARCOBRAR,250,"Saludas a los autobuseros, por ser un buen ciudadano recibes 250 euros"))
        @mazo.al_mazo(Sorpresa.new_ir_casilla(Tipo_sorpresa::IRCASILLA, tablero, 4, "Te has caido en una alcantarilla y apareces en china, te mueves 4 casillas" ))
        @mazo.al_mazo(Sorpresa.new_ir_casilla(Tipo_sorpresa::IRCASILLA,tablero,2,"tropezar y no caerte es avanzar, te mueves dos casillas"))
        @mazo.al_mazo(Sorpresa.new_carcel(Tipo_sorpresa::IRCARCEL,tablero))
        @mazo.al_mazo(Sorpresa.new_generico(Tipo_sorpresa::PORCASAHOTEL,25,"Recibes dinero por pasar por tu territorio, toma 25" ))
        @mazo.al_mazo(Sorpresa.new_generico( Tipo_sorpresa::PORCASAHOTEL,-10,"No es tu territorio ahora pagas 10"))
        @mazo.al_mazo(Sorpresa.new_generico(Tipo_sorpresa::PORJUGADOR, -12, "le pagas al resto de jugadores 12"))
        @mazo.al_mazo(Sorpresa.new_generico(Tipo_sorpresa::PORJUGADOR, 12, "Hoy es el dia de la felicidad, todos los jugadores te dan 12"))
        @mazo.al_mazo(Sorpresa.new_salir_carcel(Tipo_sorpresa::SALIRCARCEL, @mazo ))
    end
    
    def inicializar_tablero(mazo)
      @tablero = Tablero.new(5)
      @tablero.añade_casilla(Casilla.new_casilla_descanso("Parking"))
      @tablero.añade_casilla(Casilla.new_casilla_juez(5, "Juez"))
      @tablero.añade_casilla(Casilla.new_casilla_impuesto(100, "Impuesto"))
      @tablero.añade_casilla(Casilla.new_casilla_sorpresa(mazo, "Sorpresa 1"))
      @tablero.añade_casilla(Casilla.new_casilla_sorpresa(mazo, "Sorpresa 2"))
      @tablero.añade_casilla(Casilla.new_casilla_sorpresa(mazo, "Sorpresa 3"))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C1", 150, 200, 80, 45, 140)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C2", 170, 250, 90, 55, 150)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C3", 190, 300, 100, 65, 160)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C4", 210, 350, 110, 75, 170)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C5", 230, 400, 120, 85, 180)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C6", 250, 450, 130, 95, 190)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C7", 270, 500, 140, 105, 200)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C8", 290, 550, 150, 115, 210)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C9", 310, 600, 160, 125, 220)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C10", 330, 650, 170, 135, 230)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C11", 350, 700, 180, 145, 240)))
      @tablero.añade_casilla(Casilla.new_casilla_calle(Titulo_propiedad.new("C12", 370, 750, 190, 155, 250)))
      
    end
    
    def pasar_turno
      @indice_jugador_actual = (@indice_jugador_actual + 1) % @jugadores.size
    end
    
    def ranking
      @jugadores.sort
    end
    
    def salir_carcel_pagando
      return @jugador_actual.salir_carcel_pagando
    end
    
    def salir_carcel_tirando
      return @jugador_actual.salir_carcel_tirando
    end
    
=begin
def siguiente_paso
      
    end
def avanza_jugador()
      
    end
    
    
    
   
    
    def comprar
      
    end

=end
    
    
    
    
    def siguiente_paso_completado(operacion)
      @estado = @gestor_estados.siguiente_estado(@jugador_actual, @estado, operacion)
    end
    
    def vender(ip)
      return @jugador_actual.vender(ip)
    end
                
    private  :contabilizar_pasos_por_salida, :inicializar_tablero,
            :pasar_turno, :ranking, :inicializar_mazo_sorpresas
    
  end
end

File: 2ºAÑO/1ºCuatrimestre/PDOO/Civitas/Civitas/src/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author Jose
 */
public class CivitasJuego {
    
    private int indiceJugadorActual;
    private Tablero tablero;
    private ArrayList<Jugador> jugadores = new ArrayList<>();
    private EstadosJuego estado ;
    private GestorEstados gestorEstados;
    private MazoSorpresas mazo;
    
    CivitasJuego(ArrayList<String> nombres){
        for(int i = 0; i < nombres.size(); i++)
        {
            Jugador jugador = new Jugador(nombres.get(i));
            jugadores.add(jugador);
        }
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
        mazo = new MazoSorpresas(); 
        this.inicializarTablero(mazo);
        this.inicialiarMazoSorpresas(tablero);
    }
    
    /*private void avanzaJugador(){
        
    }*/
    
    /*public boolean cancelarHipoteca(int ip){
        
        return true;
    }*/
    
    /*public boolean comprar(){
        return true;
    }*/
    
    /*public boolean construirCasa(int ip){
        return this.getJugadorActual().puedeEdificarCasa(titulo);
    }
    
    public boolean construirHotel(int ip){
        return this.getJugadorActual().puedeEdificarCasa(titulo);
    }*/
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        boolean seguir=true;
                
        while(seguir)
        {
            if(tablero.getPorSalida() ==  0)
                seguir = false;
            else
                jugadorActual.pasaPorSalida();
        }
    }
    
    public boolean finalDelJuego(){
        boolean end = false;
        
        for(int i = 0; i < jugadores.size() && !end; i++)
        {
            if(jugadores.get(i).enBancarrota())
                end = true;
        }
        
        return end;
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    /*public boolean hipotecar(int ip){
        return this.getJugadorActual().hipotecar(ip);;
    }*/
    
    public String infoJugadorTexto(){
        return jugadores.get(indiceJugadorActual).toString();
    }
    
     void inicialiarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -100, "Te han multado por ir a 10km/h con el coche TORTUGA"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR,250,"Saludas a los autobuseros, por ser un buen ciudadano recibes 250 euros"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 4, "Te has caido en una alcantarilla y apareces en china, te mueves 4 casillas" ));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA,tablero,2,"tropezar y no caerte es avanzar, te mueves dos casillas"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRACARCEL,tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL,25,"Recibes dinero por pasar por tu territorio, toma 25" ));
        mazo.alMazo(new Sorpresa( TipoSorpresa.PORCASAHOTEL,-10,"No es tu territorio ahora pagas 10"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, -12, "le pagas al resto de jugadores 12"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, 12, "Hoy es el dia de la felicidad, todos los jugadores te dan 12"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo ));
    }
    
     void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(5); //se supone que puedes poner la casilla de carcel que quieras        
        tablero.añadeCasilla(new Casilla("Parking"));
        tablero.añadeCasilla(new Casilla(5, "Juez"));
        tablero.añadeCasilla(new Casilla(100f, "Impuesto"));
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa 1"));
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa 2"));
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa 3"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C1", 150f, 200f, 80f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C2", 170f, 250f, 90f, 55f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C3", 190f, 300f, 100f, 65f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C4", 210f, 350f, 110f, 75f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C5", 230f, 400f, 120f, 85f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C6", 250f, 450f, 130f, 95f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C7", 270f, 500f, 140f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C8", 290f, 550f, 150f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C9", 310f, 600f, 160f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C10", 330f, 650f, 170f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C11", 350f, 700f, 180f, 45f, 140f)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C12", 370f, 750f, 190f, 45f, 140f)));
    }
    
    private void pasarTurno(){
        indiceJugadorActual = (indiceJugadorActual + 1) % jugadores.size();     //preguntar a David
    }
    
    private ArrayList<Jugador> ranking(){
        Collections.sort(jugadores);
        return jugadores;
    }
    
    public boolean salirCarcelPagando(){
        return this.getJugadorActual().salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return this.getJugadorActual().salirCarcelTirando();
    }
    
    public Casilla getCasillaActual(){
        int casilla = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        return tablero.getCasilla(casilla);
    }
    
    /*public OperacionesJuego siguientePaso(){
        OperacionesJuego juego = OperacionesJuego.PASAR_TURNO;
        return juego;
    }*/
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(this.getJugadorActual(), estado, operacion);
    }
    
    public boolean vender(int ip){
        return this.getJugadorActual().vender(ip);
    }
    
    
}
